File: web/reviewed-memory.js
'use strict';
// Explicitly reviewed memory, not automatic extraction or execution authority.
(() => {
  const state={epoch:0,last:0,loading:false,data:null,signature:null,inspected:null};
  names['reviewed-memory']='What you choose to remember.';
  descriptions['reviewed-memory']='People, projects and decisions, connected to the evidence you reviewed.';
  const priorRender=renderContent,priorOut=signedOut;
  function reset(){state.epoch++;state.last=0;state.loading=false;state.data=null;state.signature=null;state.inspected=null;}
  signedOut=function(){reset();priorOut();};
  renderContent=function(){
    if(ui.view!=='reviewed-memory')return priorRender();
    const root=$('view-content');
    if(!root.querySelector('.reviewed-memory')){reset();root.replaceChildren();const shell=node('section','reviewed-memory');
      const tools=node('div','memory-toolbar');tools.append(button('New entity','outline',()=>act(entityDialog)),button('Propose a statement','primary',()=>act(proposalDialog)),button('Export reviewed memory','text-button',()=>act(exportMemory)));
      tools.querySelectorAll('button').forEach((b,i)=>b.id=['memory-new-entity','memory-propose','memory-export'][i]);
      const status=node('p','status-line','Reading your reviewed memory…');status.id='memory-status';status.setAttribute('role','status');
      const body=node('div');body.id='memory-records';shell.append(tools,status,body,node('p','memory-limit','Private to this access key. Acceptance records your judgement, not verified truth. Notes remain unchanged.'));root.append(shell);
    }
    root.querySelectorAll('.memory-toolbar button').forEach(b=>b.disabled=!ui.connected||Boolean(window.ALFRED_PREVIEW)||(b.id!=='memory-export'&&ui.state.role!=='owner'));
    if(ui.connected&&!state.loading&&Date.now()-state.last>2000)load();
  };
  function alive(e,g){return e===state.epoch&&g===ui.generation&&ui.csrf&&ui.view==='reviewed-memory';}
  async function load(){state.loading=true;const e=state.epoch,g=ui.generation;
    try{const data=await api('/desk/memory');if(!alive(e,g))return;state.data=data;state.last=Date.now();
      const signature=JSON.stringify(data);if(signature!==state.signature){state.signature=signature;draw(data);}
    }catch(error){if(alive(e,g))$('memory-status').textContent=errorText(error);}finally{if(alive(e,g))state.loading=false;}
  }
  function name(id){return state.data?.entities.find(x=>x.id===id)?.name||id;}
  function draw(data){const root=$('memory-records');root.replaceChildren();
    if(state.inspected&&$('detail').open){const current=data.claims.find(c=>c.id===state.inspected.id);if(!current||!current.source||current.version!==state.inspected.version){$('detail').close();$('detail-body').replaceChildren();state.inspected=null;showToast('That statement or its evidence changed. Open the current record again.');}}
    $('memory-status').textContent=data.counts.usable+' current reviewed statements · '+data.counts.proposed+' awaiting review'+(data.counts.conflicted?' · '+data.counts.conflicted+' conflicting records':'');
    const entities=node('div','memory-entities');for(const entity of data.entities){const item=node('div','memory-entity');item.append(node('strong','',entity.name),node('small','',pretty(entity.kind)+' · '+entity.id));entities.append(item);}root.append(entities);
    if(!data.claims.length)root.append(empty('Build memory deliberately','Create an entity, then propose a statement with the exact source lines. A note link alone is not a fact.'));
    for(const c of data.claims){const row=node('article','memory-claim');row.dataset.claimId=c.id;
      row.append(node('div','memory-claim-state',pretty(c.state)+(c.conflicts.length?' / conflict to resolve':'')+(!c.valid_now?' / outside validity period':'')));
      const heading=node('h2','',name(c.subject_id)+' · '+pretty(c.predicate));row.append(heading,node('p','memory-value',c.value|| (c.object_id?name(c.object_id):'Statement removed because its evidence is no longer current.')));
      if(c.source){const s=c.source;row.append(node('blockquote','memory-quote',s.quote));const source=button(s.path+' · lines '+s.start_line+'–'+s.end_line,'text-button',()=>inspect(c));source.dataset.sourceId=s.note_id;row.append(source);}
      row.append(node('small','',c.usable?'Available in the reviewed graph. No action permission is implied.':c.conflicts.length?'Competing accepted statements are excluded from the current graph.':'Not a current accepted statement.'));
      if(ui.state.role==='owner'&&['proposed','accepted','disputed'].includes(c.state)){
        const actions=node('div','memory-actions');
        for(const [op,label] of [['accept','Accept statement'],['dispute','Dispute'],['withdraw','Withdraw']]){const b=button(label,'outline',()=>act(()=>reviewDialog(c,op)));b.dataset.review=op;b.disabled=!ui.connected||Boolean(window.ALFRED_PREVIEW);actions.append(b);}
        const alternatives=data.claims.filter(x=>x.id!==c.id&&x.subject_id===c.subject_id&&x.predicate===c.predicate&&['accepted','disputed','invalidated'].includes(x.state));
        if(alternatives.length){const b=button('Replace an earlier statement','text-button',()=>reviewDialog(c,'supersede',alternatives));b.dataset.review='supersede';b.disabled=!ui.connected||Boolean(window.ALFRED_PREVIEW);actions.append(b);}row.append(actions);
      }root.append(row);
    }
    const graph=node('details','memory-graph');graph.append(node('summary','','Reviewed connections ('+data.graph.edges.length+')'));
    for(const edge of data.graph.edges)graph.append(node('p','',name(edge.source)+' → '+pretty(edge.predicate)+' → '+name(edge.target)));
    if(!data.graph.edges.length)graph.append(node('p','','Accepted, current, unconflicted entity relationships appear here. The original note graph is unchanged.'));root.append(graph);
  }
  function field(form,labelText,id,type='text'){const label=node('label','',labelText);const input=node('input');input.id=id;input.type=type;label.htmlFor=id;form.append(label,input);return input;}
  function select(form,labelText,id,options){const label=node('label','',labelText),input=node('select');label.htmlFor=id;input.id=id;for(const [v,t]of options){const o=node('option','',t);o.value=v;input.append(o);}form.append(label,input);return input;}
  function dialog(title){state.inspected=null;const body=setDialog('REVIEWED MEMORY');const h=node('h2','dialog-title',title);h.id='detail-title';body.append(h);return body;}
  async function entityDialog(){const body=dialog('Give this entity its own identity.'),form=node('form','memory-form');
    const input=field(form,'Name','memory-entity-name');input.required=true;input.maxLength=120;
    const kind=select(form,'Kind','memory-entity-kind',['person','organisation','project','asset','decision','commitment','event'].map(x=>[x,pretty(x)]));
    const submit=node('button','primary','Create entity');submit.type='submit';form.append(node('p','hint','Identical names remain separate entities. Names never grant access.'),submit);body.append(form);
    form.addEventListener('submit',e=>{e.preventDefault();act(async()=>{submit.disabled=true;try{await api('/desk/memory/entities',{id:'entity-'+crypto.randomUUID(),name:input.value,kind:kind.value});$('detail').close();await load();}finally{submit.disabled=false;}});});input.focus();
  }
  async function proposalDialog(){
    if(!state.data?.entities.length){showToast('Create an entity first.');return;}
    const e=state.epoch,g=ui.generation;const graph=await api('/desk/knowledge');if(!alive(e,g))return;
    if(!graph.nodes.length){showToast('Connect an indexed Markdown source first.');return;}
    const body=dialog('Propose a source-backed statement.'),form=node('form','memory-form');let note=null;let serial=0;
    const entities=state.data.entities;
    const subject=select(form,'About','memory-subject',entities.map(x=>[x.id,x.name+' · '+x.id]));
    const predicate=select(form,'Relationship','memory-predicate',['responsible_person','depends_on','status','scheduled_for','decision'].map(x=>[x,pretty(x)]));
    const object=select(form,'Related entity','memory-object',entities.map(x=>[x.id,x.name+' · '+x.id]));
    const value=field(form,'Statement value','memory-value');value.maxLength=400;
    function relation(){const related=['responsible_person','depends_on'].includes(predicate.value);object.disabled=!related;value.disabled=related;value.required=!related;}
    predicate.addEventListener('change',relation);relation();
    const source=select(form,'Evidence note','memory-note',graph.nodes.map(x=>[x.id,x.path]));
    const first=field(form,'First supporting line','memory-first','number'),last=field(form,'Last supporting line','memory-last','number');first.value='1';last.value='1';first.min=last.min='1';first.required=last.required=true;
    const preview=node('pre','memory-note-text','Loading source…');preview.id='memory-note-text';form.append(preview);
    const submit=node('button','primary','Save proposal for review');submit.type='submit';submit.disabled=true;form.append(node('p','hint','Nothing is accepted automatically. Select at most eight exact supporting lines. This form creates no device or account permissions.'),submit);body.append(form);
    async function read(){const ticket=++serial;note=null;submit.disabled=true;try{const n=await api('/desk/knowledge/notes/'+source.value);if(ticket!==serial||!form.isConnected)return;note=n;preview.textContent=n.body.split(/\r?\n/).map((l,i)=>(i+1)+'  '+l).join('\n');submit.disabled=false;}catch(error){if(form.isConnected)preview.textContent=errorText(error);}}
    source.addEventListener('change',read);await read();
    const request='memory-'+crypto.randomUUID();form.addEventListener('submit',ev=>{ev.preventDefault();act(async()=>{if(!note)return;submit.disabled=true;try{await api('/desk/memory/proposals',{request_id:request,subject_id:subject.value,predicate:predicate.value,object_id:object.disabled?null:object.value,value:value.disabled?null:value.value,valid_from:null,valid_until:null,evidence:{note_id:note.id,sha256:note.sha256,revision:note.revision,start_line:Number(first.value),end_line:Number(last.value)}});$('detail').close();await load();}finally{submit.disabled=false;}});});
  }
  function inspect(c){const body=dialog('The evidence behind this statement.');state.inspected={id:c.id,version:c.version};body.append(node('p','detail-summary',c.source.path),node('pre','memory-note-text',c.source.quote),node('p','hint','Source SHA-256: '+c.source.sha256),node('p','hint','Review: '+c.state+' · version '+c.version+' · reviewer '+(c.reviewer||'not reviewed')));}
  function reviewDialog(c,decision,alternatives=[]){const body=dialog(decision==='supersede'?'Replace a previous statement?':pretty(decision)+' this statement?');state.inspected={id:c.id,version:c.version};body.append(node('p','detail-summary',name(c.subject_id)+' · '+pretty(c.predicate)+' · '+(c.value||name(c.object_id))),node('blockquote','memory-quote',c.source?.quote||''));
    body.append(node('p','hint','Review records your judgement. It does not independently verify the claim or grant permission. Competing accepted values remain a visible conflict.'));
    const choices=alternatives.length?select(body,'Earlier statement to replace','memory-replaces',alternatives.map(x=>[x.id,(x.value||name(x.object_id)||'Invalidated')+' · '+x.id])):null;
    const b=button('Confirm '+decision,'primary',()=>act(async()=>{b.disabled=true;try{const prior=choices?alternatives.find(x=>x.id===choices.value):null;await api('/desk/memory/claims/'+c.id+'/review',{version:c.version,decision,replaces_id:prior?.id||null,replaces_version:prior?.version||null});$('detail').close();await load();}finally{b.disabled=false;}}));b.id='memory-review-confirm';body.append(b);
  }
  async function exportMemory(){const data=await api('/desk/memory/export');const blob=new Blob([JSON.stringify({schema:1,...data},null,2)],{type:'application/json'});const url=URL.createObjectURL(blob),link=document.createElement('a');link.href=url;link.download='ALFRED-reviewed-memory.json';link.click();setTimeout(()=>URL.revokeObjectURL(url),1000);}
  // A separate, deliberate lifecycle operation, not another execution capability.
  const pulseRender=renderContent;
  renderContent=function(){pulseRender();if(ui.view!=='pulse'||!$('routine-history')||$('pulse-history-manage'))return;const b=button('Manage retained reports','text-button',()=>act(async()=>{const plan=await api('/desk/pulse/history');const body=dialog('Retain receipts, remove older reports.');body.append(node('p','detail-summary',plan.eligible+' report summaries can be removed.'),node('p','hint','The latest 64 records, the rolling 24 hours and active runs are kept. Compact retry receipts remain. This does not change schedules or securely erase database backups.'));
    const confirm=button('Remove '+plan.eligible+' older reports','primary',()=>act(async()=>{await api('/desk/pulse/history',{fingerprint:plan.fingerprint});$('detail').close();showToast('Older report summaries removed. Retry receipts retained.');selectView('pulse');}));confirm.id='pulse-history-confirm';confirm.disabled=!plan.can_prune;body.append(confirm);}));b.id='pulse-history-manage';b.disabled=ui.state.role!=='owner'||!ui.connected||Boolean(window.ALFRED_PREVIEW);$('view-content').append(b);};
})();

File: alfred/reviewed_memory.py
"""Reviewed, actor-private statements over indexed evidence, not automated truth.

Entities are explicitly created, never merged by name. Review records a user's
judgement; it grants no device/tool authority. Source slices are reconstructed,
not duplicated. Invalidated statement values are cleared on reconciliation.
"""
from __future__ import annotations
import hashlib
import json
import secrets
from .local import Fault, exact, ident, text, timestamp, fingerprint

KINDS = ('person', 'organisation', 'project', 'asset', 'decision', 'commitment', 'event')
RELATIONS = {
    'responsible_person': {'objects': ('person',), 'single': True},
    'depends_on': {'objects': KINDS, 'single': False},
    'status': {'objects': (), 'single': True},
    'scheduled_for': {'objects': (), 'single': True},
    'decision': {'objects': (), 'single': True},
}
MAX_ENTITIES, MAX_CLAIMS = 128, 256
SCHEMA = '''
CREATE TABLE IF NOT EXISTS reviewed_memory_meta(version INTEGER NOT NULL);
INSERT INTO reviewed_memory_meta SELECT 1 WHERE NOT EXISTS(SELECT 1 FROM reviewed_memory_meta);
CREATE TABLE IF NOT EXISTS memory_entities(
 id TEXT PRIMARY KEY, scope TEXT NOT NULL, actor TEXT NOT NULL, kind TEXT NOT NULL,
 name TEXT NOT NULL, created INTEGER NOT NULL);
CREATE INDEX IF NOT EXISTS memory_entities_owner ON memory_entities(scope,actor);
CREATE TABLE IF NOT EXISTS memory_claims(
 id TEXT PRIMARY KEY, scope TEXT NOT NULL, actor TEXT NOT NULL, request_id TEXT NOT NULL,
 request_hash TEXT NOT NULL, subject_id TEXT NOT NULL, predicate TEXT NOT NULL,
 object_id TEXT, value TEXT, valid_from INTEGER, valid_until INTEGER,
 note_id TEXT NOT NULL, note_hash TEXT NOT NULL, note_revision INTEGER NOT NULL,
 first_line INTEGER NOT NULL, last_line INTEGER NOT NULL, quote_hash TEXT NOT NULL,
 state TEXT NOT NULL, version INTEGER NOT NULL, created INTEGER NOT NULL,
 reviewed INTEGER, reviewer TEXT, replaces_id TEXT,
 UNIQUE(scope,actor,request_id));
CREATE INDEX IF NOT EXISTS memory_claims_owner ON memory_claims(scope,actor);
'''


def overlap(a, b):
    """Half-open validity intervals, None means unbounded."""
    return max(a['valid_from'] or 0, b['valid_from'] or 0) < min(
        a['valid_until'] if a['valid_until'] is not None else 2**53,
        b['valid_until'] if b['valid_until'] is not None else 2**53)


class ReviewedMemory:
    def __init__(self, store):
        self.store = store
        with store.connection() as db:
            db.executescript('BEGIN IMMEDIATE;\n' + SCHEMA + '\nCOMMIT;')
            if [r[0] for r in db.execute('SELECT version FROM reviewed_memory_meta')] != [1]:
                raise Fault('unsupported_reviewed_memory_version')

    @staticmethod
    def _entity(db, p, identity):
        ident(identity)
        row = db.execute('SELECT * FROM memory_entities WHERE id=? AND scope=? AND actor=?',
                         (identity, p['scope'], p['id'])).fetchone()
        if not row:
            raise Fault('memory_entity_not_found', 404)
        return dict(row)

    @staticmethod
    def _claim(db, p, identity):
        ident(identity)
        row = db.execute('SELECT * FROM memory_claims WHERE id=? AND scope=? AND actor=?',
                         (identity, p['scope'], p['id'])).fetchone()
        if not row:
            raise Fault('memory_claim_not_found', 404)
        return dict(row)

    def _source(self, db, scope, ref):
        """Read and check current indexed source in the caller's transaction."""
        exact(ref, {'note_id', 'sha256', 'revision', 'start_line', 'end_line'})
        ident(ref['note_id'])
        if type(ref['sha256']) is not str or len(ref['sha256']) != 64 or any(c not in '0123456789abcdef' for c in ref['sha256']):
            raise Fault('invalid_memory_source_hash')
        if type(ref['revision']) is not int or ref['revision'] < 1:
            raise Fault('invalid_memory_source_revision')
        first, last = ref['start_line'], ref['end_line']
        if type(first) is not int or type(last) is not int or not 1 <= first <= last or last-first >= 8:
            raise Fault('invalid_memory_source_range')
        n = db.execute('''SELECT n.* FROM knowledge_notes n
          JOIN knowledge_sources s ON n.scope=s.scope AND n.source=s.source
          JOIN credentials c ON c.scope=n.scope AND c.id=n.source
          WHERE n.scope=? AND n.id=? AND n.status='ready' AND s.status IN ('ready','attention')
          AND c.role='source' AND c.revoked=0 AND c.expires>?''',
          (scope, ref['note_id'], self.store.now())).fetchone()
        if not n or n['sha256'] != ref['sha256'] or n['revision'] != ref['revision']:
            raise Fault('memory_source_changed', 409)
        lines = n['body'].splitlines()
        if last > len(lines):
            raise Fault('invalid_memory_source_range')
        quote = '\n'.join(lines[first-1:last])
        if not quote.strip() or len(quote) > 1400:
            raise Fault('memory_source_excerpt_capacity')
        return {'note_id':n['id'], 'path':n['path'], 'title':n['title'], 'sha256':n['sha256'],
                'revision':n['revision'], 'start_line':first, 'end_line':last, 'quote':quote,
                'quote_hash':hashlib.sha256(quote.encode()).hexdigest()}

    @staticmethod
    def _ref(c):
        return {'note_id':c['note_id'], 'sha256':c['note_hash'], 'revision':c['note_revision'],
                'start_line':c['first_line'], 'end_line':c['last_line']}

    def _reconcile(self, db, p):
        for row in db.execute("SELECT * FROM memory_claims WHERE scope=? AND actor=? AND state!='invalidated' AND (value IS NOT NULL OR object_id IS NOT NULL)",
                              (p['scope'], p['id'])).fetchall():
            c=dict(row)
            try:
                source=self._source(db,p['scope'],self._ref(c))
                valid=source['quote_hash']==c['quote_hash']
            except Fault:
                valid=False
            if not valid:
                db.execute("UPDATE memory_claims SET state='invalidated',value=NULL,object_id=NULL,version=version+1 WHERE id=?",(c['id'],))
                self.store.log(db,p['scope'],p['id'],'memory.invalidated',c['id'])

    def create_entity(self, bearer, body):
        exact(body, {'id','kind','name'});ident(body['id']);text(body['name'],120)
        if type(body['kind']) is not str or body['kind'] not in KINDS or not body['name'].strip():
            raise Fault('invalid_memory_entity')
        with self.store.transaction() as db:
            p=self.store.authenticate(db,bearer,{'owner'})
            existing=db.execute('SELECT * FROM memory_entities WHERE id=?',(body['id'],)).fetchone()
            if existing:
                if (existing['scope'],existing['actor'],existing['kind'],existing['name'])!=(p['scope'],p['id'],body['kind'],body['name'].strip()):
                    raise Fault('memory_entity_id_unavailable',409)
                return {'id':body['id'],'reused':True}
            if db.execute('SELECT count(*) FROM memory_entities WHERE actor=?',(p['id'],)).fetchone()[0]>=MAX_ENTITIES:
                raise Fault('memory_entity_capacity',409)
            db.execute('INSERT INTO memory_entities VALUES (?,?,?,?,?,?)',
                       (body['id'],p['scope'],p['id'],body['kind'],body['name'].strip(),self.store.now()))
            self.store.log(db,p['scope'],p['id'],'memory.entity_created',body['id'])
        return {'id':body['id'],'reused':False}

    def propose(self, bearer, body):
        exact(body, {'request_id','subject_id','predicate','object_id','value','valid_from','valid_until','evidence'})
        ident(body['request_id']);ident(body['subject_id'])
        if type(body['predicate']) is not str or body['predicate'] not in RELATIONS:
            raise Fault('invalid_memory_predicate')
        relation=RELATIONS[body['predicate']]
        if relation['objects']:
            ident(body['object_id'])
            if body['value'] is not None:raise Fault('memory_object_not_text')
        else:
            if body['object_id'] is not None:raise Fault('memory_value_not_entity')
            text(body['value'],400)
            if not body['value'].strip():raise Fault('invalid_memory_value')
        for key in ('valid_from','valid_until'):
            if body[key] is not None:timestamp(body[key])
        if body['valid_until'] is not None and body['valid_until'] <= (body['valid_from'] or 0):
            raise Fault('invalid_memory_validity')
        digest=fingerprint(body)
        with self.store.transaction() as db:
            p=self.store.authenticate(db,bearer,{'owner'});self._reconcile(db,p)
            old=db.execute('SELECT id,request_hash,state FROM memory_claims WHERE scope=? AND actor=? AND request_id=?',
                           (p['scope'],p['id'],body['request_id'])).fetchone()
            if old:
                if old['request_hash']!=digest:raise Fault('memory_request_collision',409)
                return {'id':old['id'],'state':old['state'],'reused':True}
            self._entity(db,p,body['subject_id'])
            if body['object_id']:
                other=self._entity(db,p,body['object_id'])
                if other['kind'] not in relation['objects']:raise Fault('memory_object_kind')
                if body['subject_id']==body['object_id']:raise Fault('memory_self_relation')
            source=self._source(db,p['scope'],body['evidence'])
            if db.execute('SELECT count(*) FROM memory_claims WHERE actor=?',(p['id'],)).fetchone()[0]>=MAX_CLAIMS:
                raise Fault('memory_claim_capacity',409)
            cid='claim-'+secrets.token_hex(12)
            db.execute('INSERT INTO memory_claims VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)',
                       (cid,p['scope'],p['id'],body['request_id'],digest,body['subject_id'],body['predicate'],
                        body['object_id'],body['value'].strip() if body['value'] else None,body['valid_from'],body['valid_until'],
                        source['note_id'],source['sha256'],source['revision'],source['start_line'],source['end_line'],
                        source['quote_hash'],'proposed',1,self.store.now(),None,None,None))
            self.store.log(db,p['scope'],p['id'],'memory.proposed',cid)
        return {'id':cid,'state':'proposed','reused':False}

    def review(self, bearer, identity, body):
        exact(body, {'version','decision','replaces_id','replaces_version'})
        if type(body['version']) is not int or body['version'] < 1 or body['decision'] not in ('accept','dispute','withdraw','supersede'):
            raise Fault('invalid_memory_review')
        if body['decision']!='supersede' and (body['replaces_id'] is not None or body['replaces_version'] is not None):
            raise Fault('unexpected_memory_replacement')
        with self.store.transaction() as db:
            p=self.store.authenticate(db,bearer,{'owner'});self._reconcile(db,p)
            c=self._claim(db,p,identity)
            if c['state']=='invalidated':raise Fault('memory_source_changed',409)
            if c['version']!=body['version']:raise Fault('memory_review_changed',409)
            if c['state'] not in ('proposed','accepted','disputed'):raise Fault('memory_review_closed',409)
            self._source(db,p['scope'],self._ref(c))
            replacement=c['replaces_id']
            if body['decision']=='supersede':
                if c['replaces_id'] is not None:raise Fault('memory_replacement_already_recorded',409)
                if type(body['replaces_version']) is not int:raise Fault('invalid_memory_replacement')
                prior=self._claim(db,p,body['replaces_id'])
                if prior['id']==c['id'] or prior['state'] not in ('accepted','disputed','invalidated') or prior['version']!=body['replaces_version']:
                    raise Fault('memory_replacement_changed',409)
                if (prior['subject_id'],prior['predicate'])!=(c['subject_id'],c['predicate']):raise Fault('memory_replacement_mismatch')
                replacement=prior['id']
                db.execute("UPDATE memory_claims SET state='superseded',version=version+1,reviewed=?,reviewer=? WHERE id=?",
                           (self.store.now(),p['id'],replacement))
                self.store.log(db,p['scope'],p['id'],'memory.superseded',replacement)
            state={'accept':'accepted','supersede':'accepted','dispute':'disputed','withdraw':'withdrawn'}[body['decision']]
            db.execute('UPDATE memory_claims SET state=?,version=version+1,reviewed=?,reviewer=?,replaces_id=? WHERE id=?',
                       (state,self.store.now(),p['id'],replacement,c['id']))
            self.store.log(db,p['scope'],p['id'],'memory.'+state,c['id'])
        return self.view(bearer)

    def view(self, bearer):
        with self.store.transaction() as db:
            p=self.store.authenticate(db,bearer,{'owner','reader'});self._reconcile(db,p)
            entities=[dict(r) for r in db.execute('SELECT id,kind,name,created FROM memory_entities WHERE scope=? AND actor=? ORDER BY created,id',(p['scope'],p['id']))]
            claims=[];now=self.store.now()
            for row in db.execute('SELECT * FROM memory_claims WHERE scope=? AND actor=? ORDER BY created DESC,id',(p['scope'],p['id'])):
                c=dict(row)
                c.pop('request_hash');c.pop('request_id');c.pop('scope');c.pop('actor')
                c['source']=None
                if c['state']!='invalidated' and (c['value'] is not None or c['object_id'] is not None):
                    try:c['source']=self._source(db,p['scope'],self._ref(c))
                    except Fault:pass
                c['valid_now']=(c['valid_from'] is None or c['valid_from']<=now) and (c['valid_until'] is None or c['valid_until']>now)
                c['conflicts']=[];claims.append(c)
            for c in claims:
                if c['state']!='accepted' or not RELATIONS[c['predicate']]['single']:continue
                for other in claims:
                    if other['id']!=c['id'] and other['state']=='accepted' and (other['subject_id'],other['predicate'])==(c['subject_id'],c['predicate']) and (other['object_id'],other['value'])!=(c['object_id'],c['value']) and overlap(c,other):
                        c['conflicts'].append(other['id'])
            for c in claims:
                c['usable']=c['state']=='accepted' and c['valid_now'] and bool(c['source']) and not c['conflicts']
            return {'entities':entities,'claims':claims,'scope':p['scope'],'actor_private':True,
                    'basis':'user_reviewed_statements_not_verified_facts','model_extraction':False,
                    'counts':{'proposed':sum(c['state']=='proposed' for c in claims),'usable':sum(c['usable'] for c in claims),
                              'conflicted':sum(bool(c['conflicts']) for c in claims)},
                    'graph':{'nodes':entities,'edges':[{'source':c['subject_id'],'target':c['object_id'],'predicate':c['predicate'],'claim_id':c['id']} for c in claims if c['usable'] and c['object_id']]},
                    'limits':{'entities':MAX_ENTITIES,'claims':MAX_CLAIMS},'authority_granted':False}
